lib/resty/dns/resolver.lua

```diff
@@ -844,56 +844,57 @@ function _M.query(self, qname, opts)
     -- local cjson = require "cjson"
     -- print("query: ", cjson.encode(concat(query, "")))
 
+    local ok
     local retrans = self.retrans
 
     -- print("retrans: ", retrans)
 
     for i = 1, retrans do
         local sock = pick_sock(self, socks)
 
-        local ok, err = sock:send(query)
+        ok, err = sock:send(query)
         if not ok then
             local server = _get_cur_server(self)
-            return nil, "failed to send request to UDP server "
+            err = "failed to send request to UDP server "
                 .. concat(server, ":") .. ": " .. err
-        end
-
-        local buf, err
-
-        for j = 1, 128 do
-            buf, err = sock:receive(4096)
 
-            if err then
-                break
-            end
+        else
+            local buf
+
+            for _ = 1, 128 do
+                buf, err = sock:receive(4096)
+                if err then
+                    local server = _get_cur_server(self)
+                    err = "failed to receive reply from UDP server "
+                        .. concat(server, ":") .. ": " .. err
+                    break
+                end
 
-            if buf then
-                local answers
-                answers, err = parse_response(buf, id, opts)
-                if not answers then
+                if buf then
+                    local answers
+                    answers, err = parse_response(buf, id, opts)
                     if err == "truncated" then
-                        return _tcp_query(self, query, id, opts)
+                        answers, err = _tcp_query(self, query, id, opts)
                     end
 
-                    if err ~= "id mismatch" then
-                        return nil, err
+                    if err and err ~= "id mismatch" then
+                        break
                     end
 
-                    -- retry receiving when err == "id mismatch"
-                else
-                    return answers
+                    if answers then
+                        return answers, nil, tries
+                    end
                 end
+                -- only here in case of an "id mismatch"
             end
         end
 
-        if err ~= "timeout" or i == retrans then
-            local server = _get_cur_server(self)
-            return nil, "failed to receive reply from UDP server "
-                .. concat(server, ":") .. ": " .. err
+        if tries then
+          tries[i] = err
```

Comment:
(Such tables that are fed by users need to be `nil` terminated before returning them so that users can easily iterate over them)